src/workerd/util/sqlite.c++

```diff
@@ -438,16 +447,62 @@ kj::Own<sqlite3_stmt> SqliteDatabase::prepareSql(
           } else {
             SQLITE_CALL_FAILED("sqlite3_step()", err);
           }
-
-          // Reduce `sqlCode` to include only what we haven't already executed.
-          sqlCode = kj::StringPtr(tail, sqlCode.end());
-          continue;
         }
-        break;
+
+        // Now set statementResult as the "result" to be either returned or step()-ed through
+        result = statementResult;
+      }
+
+      // Reduce `sqlCode` to include only what we haven't already executed.
+      sqlCode = kj::StringPtr(tail, sqlCode.end());
+    }
+
+    // After processing the entire input, throw if we never got a single result
+    SQLITE_REQUIRE(result != nullptr, "SQL code did not contain a complete statement.", sqlCode);
+    return ownSqlite(result);
+  }
+}
+
+int SqliteDatabase::ingestSql(Regulator& regulator, kj::StringPtr sqlCode) {
+  // Set up the regulator that will be used for authorizer callbacks while preparing this
+  // statement.
+  KJ_ASSERT(currentRegulator == nullptr, "recursive prepareSql()?");
+  KJ_DEFER(currentRegulator = nullptr);
+  currentRegulator = regulator;
+  const char* start = sqlCode.begin();
+  const char* tail = start;
+
+  SQLITE_REQUIRE(sqlite3_get_autocommit(db) == 0, "sql.ingest must be run inside a transaction");
```

Comment:
Some debate here whether this should be a real constraint of the new API or just a convention we suggest users follow.

Note that if we remove this constraint we must add the code to call the `onWriteCallback` to set up an implicit transaction if not already present (see line 431 above).

Reply:
I don't think requiring a transaction is warranted:

* It seems like the primary intended use case for ingest() is for uploading a database dump in a streaming way. But a database which is being restored from a dump is almost certainly not serving live traffic. So while it is arguably a good idea to wrap it in a transaction, it probably doesn't matter for this use case.
* We don't know what the other use cases might be, but if there are others, I am not at all convinced that wrapping in a transaction is always the right thing to do. Forcing the developer to do it feels like a nanny-state type of thing to do.
* I think developers are plenty capable of reasoning for themselves about transactions. Or if they aren't, forgetting to apply one to ingest() is likely not going to be their biggest mistake.
* I think requiring a transaction makes it harder to explain what ingest() does. The explanation should be as simple as: "It's like exec(), except that it only parses complete statements ending in a semicolon. Any incomplete statement at the end is returned to you."

Reply:
> It's like exec(), except that it only parses complete statements ending in a semicolon. Any incomplete statement at the end is returned to you.

Thanks for this, that's a better API and removes the need to even talk about transactions.